marklogic-data-hub/src/main/java/com/marklogic/hub/impl/FlowManagerImpl.java

```diff
@@ -184,18 +223,102 @@ public boolean isFlowExisted(String flowName) {
         return false;
     }
 
+    @Deprecated
+    @Override
     public Map<String, Step> getSteps(Flow flow) {
         return flow.getSteps();
     }
 
+    @Override
+    @Deprecated
     public Step getStep(Flow flow, String stepNum) {
         return flow.getStep(stepNum);
     }
 
+    @Override
+    @Deprecated
     public void setSteps(Flow flow, Map<String, Step> stepMap) {
         flow.setSteps(stepMap);
     }
 
+    /**
+     * For certain step types, when the step is deleted, we need to do some cleanup of artifacts on disk and in the
+     * staging and final databases.
+     *
+     * @param flow
+     * @param removedStep
+     */
+    protected void deleteRelatedStepArtifacts(Flow flow, Step removedStep) {
+        if (removedStep == null) {
+            return;
+        }
+
+        if (removedStep.isMappingStep() && !stepIsReferencedByAFlow(removedStep.getName(), StepDefinition.StepDefinitionType.MAPPING)) {
+            deleteMappingArtifacts(flow, removedStep);
+        } else if (removedStep.isCustomStep() && !stepIsReferencedByAFlow(removedStep.getName(), StepDefinition.StepDefinitionType.CUSTOM)) {
+            deleteStepDefinitionArtifacts(removedStep);
+        }
+    }
+
+    /**
+     * @param stepName
+     * @param stepType
+     * @return true if any flow has a step with the given name and type
+     */
+    protected boolean stepIsReferencedByAFlow(String stepName, StepDefinition.StepDefinitionType stepType) {
+        for (Flow flow : getFlows()) {
+            for (Step step : flow.getSteps().values()) {
+                if (stepType.equals(step.getStepDefinitionType()) && stepName.equals(step.getName())) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    protected void deleteMappingArtifacts(Flow flow, Step removedStep) {
+        logger.info("Deleting mapping as it's no longer referenced by any flows: " + removedStep.getName());
+        final String mappingName = flow.getName() + "-" + removedStep.getName();
+        mappingManager.deleteMapping(mappingName);
+        deleteDocumentsInDirectory(format("/mappings/%s/", mappingName));
+    }
+
+    protected void deleteStepDefinitionArtifacts(Step removedStep) {
+        logger.info("Deleting custom step as it's no longer referenced by any flows: " + removedStep.getName() +
+            ". The module associated with this step will not be deleted in case other modules refer to it.");
+        StepDefinition stepDef = StepDefinition.create(removedStep.getName(), StepDefinition.StepDefinitionType.CUSTOM);
+        stepDefinitionManager.deleteStepDefinition(stepDef);
+        deleteDocumentsInDirectory(format("/step-definitions/%s/%s/", stepDef.getType().toString(), stepDef.getName()));
+    }
+
+    protected void deleteDocumentsInDirectory(String directory) {
```

Comment:
We can probably reuse `dataHubService.deleteDocument(String uri, DatabaseKind databaseKind)`

Reply:
What do you mean by multiple documents? Wouldn't there only be a single document for mapping/step-def for a step that is being deleted?

The URI's for step-defs and mappings obey a set standard. So we know the URI in advance and can delete them using one or more calls to delete the URIs.

Reply:
A mapping directory can have multiple documents in it, one for each version. I think a step definition can only have a single document, but it seems safer for the future to just delete the directory since we know everything in the directory should be deleted.

Reply:
Makes sense.